### PR TITLE
Add chart tests

### PR DESCRIPTION
**Pull request by corest**
Towards https://github.com/giantswarm/giantswarm/issues/4138

- basic test for chart validation
- testing availability of the metrics

### CODE REVIEW

File: integration/test/metrics/main_test.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+	"github.com/giantswarm/micrologger"
+	"github.com/spf13/afero"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/kubernetes-metrics-server/integration/templates"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/rest"
+)
```
Comment (pipo02mix) [inline integration/test/metrics/main_test.go]: Nitpick. Move externals in the previous block

**Review by pipo02mix — APPROVED** [review]
LGTM besides the nitpick

File: integration/test/metrics/metrics_test.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	err = checkMetricsAvailability()
+	if err != nil {
+		t.Fatalf("could not get metrics: %v", err)
+	}
+}
+
+func checkMetricsAvailability() error {
+
+	restConfig := h.RestConfig()
+	restClient, err := rest.RESTClientFor(restConfig)
```
Comment (fgimenez) [inline integration/test/metrics/metrics_test.go]: No need to instantiate a new REST client, you have one with `h.K8sClient().CoreV1().RESTClient()`

File: integration/test/metrics/metrics_test.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	stream, err := restClient.Get().RequestURI(metricsAPIEndpoint).Stream()
+	if err != nil {
+		return err
+	}
+	defer stream.Close()
```
Comment (fgimenez) [inline integration/test/metrics/metrics_test.go]: `defer` here is redundant here.

File: integration/test/metrics/metrics_test.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	metricsAPIEndpoint = "/apis/metrics.k8s.io/v1beta1"
+)
+
+// TestMetrics ensures that deployed metrics-server chart exposes node-metrics
+// via Kubernetes API extension.
+func TestMetrics(t *testing.T) {
+	ctx := context.Background()
+
+	// Install resource
+	err := r.InstallResource(chartName, templates.MetricsServerValues, fmt.Sprintf("%s-%s", env.CircleSHA(), testName))
```
Comment (fgimenez) [inline integration/test/metrics/metrics_test.go:31]: A teardown for this resource is missing.

File: integration/test/metrics/metrics_test.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	// Wait for deployed status
+	err = r.WaitForStatus(chartName, "DEPLOYED")
+	if err != nil {
+		t.Fatalf("timeout waiting for deployed resource: %v", err)
+	}
+
+	// Wait 1 minute for metrics become available
+	l.LogCtx(ctx, "level", "info", "message", "waiting 1 minute for metrics become available")
```
Comment (fgimenez) [inline integration/test/metrics/metrics_test.go]: The common pattern is using a backoff when we need to wait for something to happen, maybe the metrics endpoint is available before the minute has passed, maybe it is a little bit later. See https://github.com/giantswarm/e2etests/blob/master/update/update.go#L57-L74 for an example.

**Review by fgimenez — COMMENTED** [review]
Direction looks very good.

File: integration/test/metrics/metrics_test.go (hunk 4 above)
Comment (corest) [reply integration/test/metrics/metrics_test.go:31]: Isn't that handled in main test setup?

Comment (fgimenez) [reply integration/test/metrics/metrics_test.go:31]: We usually add a teardown file on the setup package to remove common resources created during setup, see https://github.com/giantswarm/aws-operator/blob/master/integration/setup/teardown.go. If one test creates specific resources it should take care of removing them once it has finished, if not those would be left behind.

File: integration/test/metrics/metrics_test.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	}
+
+	// Check metrics availability
+	err = checkMetricsAvailability(ctx)
+	if err != nil {
+		t.Fatalf("could not get metrics: %v", err)
+	}
+
+	// Delete release
+	err = helmClient.DeleteRelease(chartName, helm.DeletePurge(true))
```
Comment (corest) [inline integration/test/metrics/metrics_test.go:49]: @fgimenez I've added release deletion here as it basically no that super advanced test

Comment (fgimenez) [reply integration/test/metrics/metrics_test.go:49]: Nice, thanks!

**Review by fgimenez — APPROVED** [review]
LGTM, maybe you could also check the delete error on teardown, but not a blocker.

**Pull request merged**